writer.go

```diff
@@ -335,6 +338,65 @@ func NewMediaPlaylist(winsize uint, capacity uint) (*MediaPlaylist, error) {
 	return p, nil
 }
 
+// InsertSegments allows the insertion of one or multiple MediaSegments into the MediaPlaylist.
+// seqID is the sequence ID at which the new segments should be inserted to
+// This operation does reset playlist cache.
+func (p *MediaPlaylist) InsertSegments(segments []*MediaSegment, seqID uint64) error {
+	if len(segments) == 0 {
+		return ErrPlaylistEmpty
+	}
+
+	// Determine the index where the new segments should be inserted
+	var insertIndex = 0
+	switch {
+	case seqID >= uint64(len(p.Segments)):
+		insertIndex = len(p.Segments)
+	case seqID != 0 && seqID < uint64(len(p.Segments)):
+		insertIndex = int(seqID) - 1
+	}
+
+	// Create copies of the segments to be inserted
+	localSegments := make([]*MediaSegment, len(segments), len(segments)+len(p.Segments[insertIndex:]))
+	copy(localSegments, segments)
+
+	// Adjust values
+	adjustment := uint(len(segments))
+	p.count += adjustment
+	p.capacity += adjustment
+	p.tail = p.count
```

Comment:
In general such adjustments should happen at the end of the function, when all related operations succeeded, though there are no errors thrown, but in case of panic and recovering from it the later use of the same `MediaPlaylist` instance will be impossible.